Firestore/core/src/firebase/firestore/firestore_version.cc

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/include/firebase/firestore/firestore_version.h"
+
+#ifndef FIRFirestore_VERSION
+#error \
+    // NOLINTNEXTLINE(whitespace/line_length)
+    "FIRFirestore_VERSION is not defined: add -DFIRFirestore_VERSION=... to the build invocation"
+#endif
+
+namespace firebase {
+namespace firestore {
+
+// The following two macros supply the incantation so that the C
+// preprocessor does not try to parse the version as a floating
+// point number. See
+// https://www.guyrutenberg.com/2008/12/20/expanding-macros-into-string-constants-in-c/
+#define STR(x) STR_EXPAND(x)
+#define STR_EXPAND(x) #x
+
+extern "C" const char* const firestore_version_string =
```

Comment:
We shouldn't have more than one place defining such constants. Ideally the existing version string could point at this one.

Also, unlike Objective-C it seems like this could have C++ linkage, no?

Reply:
Done.

Regarding dependency between Objective-C variable and C++ variable:
* I _think_ that if both are `extern`, they are still not affected by static initialization order fiasco, because `kFirestoreVersionString` should be subject to [constant initialization](https://en.cppreference.com/w/cpp/language/constant_initialization), so it's basically compile-time initialized and guaranteed to be available by the time dynamic initialization of `FIRFirestoreVersionString` runs (or at least that would be so under C++11 rules. I don't know if the interplay between languages may affect this). Please double-check my reasoning.
* Another alternative would be to make `kFirestoreVersionString` have internal linkage. This is more in line with other Firebase products, which use internal linkage for version strings. One downside is that the macros used to build the string would be unnecessarily exposed to anyone including the file.

Reply:
.mm and .cc files get C++ rules for linkage.

We can certainly make the C++ version conform to the C++ product rules and just alias that with the Objective-C constant.